Realm/ObjectServerTests/RealmServer.swift

```diff
@@ -0,0 +1,780 @@
+        }
+    }
+
+    typealias AppId = String
+    
+    /// Create a new server app
+    @objc func createApp() throws -> AppId {
+        guard let session = session else {
+            throw URLError(.unknown)
+        }
+
+        let (info, appCreationError) = session.apps.post(["name": "test"])
+        guard let appInfo = info as? [String: Any],
+              let clientAppId = appInfo["client_app_id"] as? String,
+              let appId = appInfo["_id"] as? String else {
+            throw appCreationError ?? URLError(.badServerResponse)
+        }
+
+        let app = session.apps[appId]
+        let group = DispatchGroup()
+
+        app.authProviders.post(on: group, ["type": "anon-user"], { if let error = $1 {
+            XCTFail(error.localizedDescription)
+        }})
+        app.authProviders.post(on: group,
+            [
+                "type": "local-userpass",
+                "config": [
+                    "emailConfirmationUrl": "http://foo.com",
+                    "resetPasswordUrl": "http://foo.com",
+                    "confirmEmailSubject": "Hi",
+                    "resetPasswordSubject": "Bye",
+                    "autoConfirm": true
+                ]
+            ])  { if let error = $1 {
+            XCTFail(error.localizedDescription)
+        }}
+
+        app.authProviders.get(on: group) { any, error in
+            guard let authProviders = any as? [[String: Any]] else {
+                return XCTFail()
+            }
+
+            for provider in authProviders {
+                if provider["type"] as? String == "api-key" {
+                    return app.authProviders[provider["_id"] as! String].enable.put(on: group) { if let error = $1 {
+                        XCTFail(error.localizedDescription)
+                    }}
+                }
+            }
+        }
+        
+        let (_, _) = app.secrets.post([
+            "name": "BackingDB_uri",
+            "value": "mongodb://localhost:26000"
+        ])
+
+        let (serviceResponse, serviceCreationError) = app.services.post([
+            "name": "mongodb1",
+            "type": "mongodb",
+            "config": [
+                "uri": "mongodb://localhost:26000",
+                "sync": [
+                    "state": "enabled",
+                    "database_name": "test_data",
+                    "partition": [
+                        "key": "realm_id",
+                        "type": "string",
+                        "permissions": [
+                            "read": true,
+                            "write": true
+                        ]
+                    ]
+                ]
+            ]
+        ])
+
+        guard serviceCreationError == nil,
+              let serviceId = (serviceResponse as? [String: Any])?["_id"] as? String else {
+            throw serviceCreationError ?? URLError(.badServerResponse)
+        }
+        
+        let dogRule: [String : Any] = [
```

Comment:
Obviously you don't need to do this now, but I think it'd be pretty easy to create these programmatically from an ObjectSchema and that'd simplify adding new types in the future.